File: java/prototype.java
import java.util.ArrayList;
import java.util.List;

// Interfaz Prototype
interface iPhone extends Cloneable {
    iPhone clone();
    void mostrar();
}

// Clase concreta que implementa el prototipo
class iPhoneModel implements iPhone {
    private String model;
    private int storage;
    private String color;
    private String iosVersion;

    public iPhoneModel(String model, int storage, String color, String iosVersion) {
        this.model = model;
        this.storage = storage;
        this.color = color;
        this.iosVersion = iosVersion;
    }

    @Override
    public iPhone clone() {
        return new iPhoneModel(model, storage, color, iosVersion);
    }

    @Override
    public void mostrar() {
        System.out.println("iPhone - Model: " + model + ", Storage: " + storage + "GB, Color: " + color + ", iOS Version: " + iosVersion);
    }
}

// Clase principal
public class prototype {
    public static void main(String[] args) {
        // Prototipo base de iPhone
        iPhoneModel iphonePrototype = new iPhoneModel("iPhone 14", 128, "Silver", "iOS 16");

        // Creación de nuevos iPhones a partir del prototipo
        iPhone iphone1 = iphonePrototype.clone();
        ((iPhoneModel) iphone1).mostrar();

        iPhone iphone2 = iphonePrototype.clone();
        ((iPhoneModel) iphone2).mostrar();

        iPhone iphone3 = iphonePrototype.clone();
        ((iPhoneModel) iphone3).mostrar();
    }
}

File: prototype.py
import copy
from abc import ABC, abstractmethod

class iPhone(ABC):
    @abstractmethod
    def clone(self):
        pass

    def __str__(self):
        return f"Model: {self.model}, Storage: {self.storage}GB, Color: {self.color}, iOS Version: {self.ios_version}"
    
class iPhoneModel(iPhone):
    def __init__(self, model, storage, color, ios_version):
        self.model = model
        self.storage = storage
        self.color = color
        self.ios_version = ios_version

    def clone(self):
        return copy.deepcopy(self)
        

    def __str__(self):
        return f"iPhone - Model: {self.model}, Storage: {self.storage}GB, Color: {self.color}, iOS Version: {self.ios_version}"

# Prototipo base de iPhone
iphone_prototype = iPhoneModel("iPhone 14", 128, "Silver", "iOS 16")

# Creación de nuevos iPhones a partir del prototipo
iphone1 = iphone_prototype.clone()
iphone1.model = "iPhone 15"
iphone1.color = "Gold"
iphone1.ios_version = "iOS 17"

iphone2 = iphone_prototype.clone()
iphone2.model = "iPhone 15 Pro"
iphone2.storage = 256
iphone2.color = "Titanium"

iphone3 = iphone_prototype.clone()
iphone3.model = "iPhone SE"
iphone3.storage = 64
iphone3.color = "Red"
iphone3.ios_version = "iOS 16"

# Imprimir resultados
print(iphone1)
print(iphone2)
print(iphone3)
